### PR TITLE
Per wallet application

### PR DESCRIPTION
**Pull request by eprbell**
Full implementation of per-wallet transfer analysis including comprehensive unit tests. This covers all the corner cases I was able to find. If you find a corner case that is not covered by a unit tests do let me know. Note that unit tests use [Go style, table-driven format](https://lorenzopeppoloni.com/tabledriventestspy/): this format is far superior to custom code and I think we should be using it for all new unit tests. Essentially it makes tests very easy to read and mantain: even non programmers could potentially read them and contribute ideas for new ones.

The code is essentially ready for review (except for a few minor lint issues).

### CODE REVIEW

File: src/rp2/accounting_engine.py [hunk 1]
```diff
@@ -204,21 +204,24 @@ def get_acquired_lot_for_taxable_event(
...
+            raise AcquiredLotsExhaustedException()
+        if acquired_lot_and_index.acquired_lot != self.__acquired_lot_list[acquired_lot_and_index.index]:
+            raise RP2RuntimeError("Internal error: acquired_lot incongruence in accounting logic")
+        method = self._get_accounting_method(taxable_event.timestamp.year)
+        lot_candidates: Optional[AbstractAcquiredLotCandidates] = self.__years_2_lot_candidates.find_max_value_less_than(taxable_event.timestamp.year)
+        # lot_candidates is 1:1 with acquired_lot_and_index, should always be True
+        if not lot_candidates:
+            raise RP2RuntimeError("Internal error: no lot candidates found for year")
+        lot_candidates.set_to_index(acquired_lot_and_index.index)
+        acquired_lot_and_amount: Optional[AcquiredLotAndAmount] = method.seek_non_exhausted_acquired_lot(lot_candidates, new_taxable_event_amount)
```
Comment (macanudo527) [inline src/rp2/accounting_engine.py:218]: Doesn't this need `taxable_event`?
```suggestion
        acquired_lot_and_amount: Optional[AcquiredLotAndAmount] = method.seek_non_exhausted_acquired_lot(lot_candidates, taxable_event, new_taxable_event_amount)
```

Comment (macanudo527) [reply src/rp2/accounting_engine.py:218]: Nevermind, I was looking at stale code.

File: src/rp2/in_transaction.py [hunk 2]
```diff
@@ -63,8 +72,22 @@ def __init__(
...
+
+        # After a transfer, the per-wallet application model is populated with artificial InTransactions that model the "to" part
+        # of the transfer (these artificial InTransactions are not present in the universal application model). These fields are
+        # only populated in the artificial InTransactions of the per-wallet application model field and describes where the funds
+        # come from (due to the transfer) and go to (after the transfer).
         self.__from_lot: Optional[InTransaction] = InTransaction.type_check("from_lot", from_lot) if from_lot is not None else None
-        self.__to_lots: Dict[str, List[InTransaction]] = {}
+        self.__to_lots: Dict[Account, List[InTransaction]] = {}
+
+        # This field is also used only in the artificial InTransactions of the per-wallet application model. If captures all the
```
Comment (macanudo527) [inline src/rp2/in_transaction.py:83]:
```suggestion
        # This field is also used only in the artificial InTransactions of the per-wallet application model. It captures all the
```

File: src/rp2/in_transaction.py [hunk 3]
```diff
@@ -63,8 +72,22 @@ def __init__(
             self.__crypto_in = configuration.type_check_positive_decimal("crypto_in", crypto_in, non_zero=True)
         self.__crypto_fee: RP2Decimal = configuration.type_check_positive_decimal("crypto_fee", crypto_fee) if crypto_fee else ZERO
         self.__fiat_fee: RP2Decimal = configuration.type_check_positive_decimal("fiat_fee", fiat_fee) if fiat_fee else ZERO
+
+        # After a transfer, the per-wallet application model is populated with artificial InTransactions that model the "to" part
+        # of the transfer (these artificial InTransactions are not present in the universal application model). These fields are
+        # only populated in the artificial InTransactions of the per-wallet application model field and describes where the funds
+        # come from (due to the transfer) and go to (after the transfer).
```
Comment (macanudo527) [inline src/rp2/in_transaction.py:79]: A bit of a nitpick, but technically I think this comment should go below `self.__from_lot`

**Comment by macanudo527** [comment]
@eprbell I've been reading this over and over and trying to digest it all. So far it looks pretty good. If you have the time can you update the wiki? Specifically the [section about `InTransaction`](https://github.com/eprbell/rp2/wiki/Adding-Per%E2%80%90Wallet-Application-to-RP2#:~:text=__parent%20is%20populated,per%2Dwallet%20analysis).) so we have a better understanding of the plan for this code.

**Comment by eprbell** [comment]
> @eprbell I've been reading this over and over and trying to digest it all. So far it looks pretty good. If you have the time can you update the wiki? Specifically the [section about `InTransaction`](https://github.com/eprbell/rp2/wiki/Adding-Per%E2%80%90Wallet-Application-to-RP2#:~:text=__parent%20is%20populated,per%2Dwallet%20analysis).) so we have a better understanding of the plan for this code.

Right. I'm currently working on universal application, but when I get some time I'll update the Wiki, because it's a bit outdated compared to the current implementation.

**Comment by eprbell** [comment]
Fixed a bug in fee handling during transfer analysis. Also added check of actual amounts in tests.

**Comment by macanudo527** [comment]
@eprbell I know it is the holidays, but I thought I would get the ball rolling on this again. Is this about ready? Are there any major changes needed? I don't have that much time, but I can try to take a look at it here and there, so we can get this out for 2025 tax filing.

**Comment by eprbell** [comment]
I can confirm that unfortunately I won't have much time to work on this (and on RP2 in general) for the foreseeable future, due to heavy workload at my job. I don't think I'll be able to get any free cycles before Tax Day.
